docs/keyed_services.md

```diff
@@ -0,0 +1,153 @@
+    JNIEnv* env,
+    const base::android::JavaParamRef<jobject>& profile_android) {
+  auto* profile = Profile::FromJavaObject(profile_android);
+  auto pending =
+      FakeServiceFactory::GetInstance()->GetRemoteService(profile);
+  return pending.PassPipe().release().value()
+}
+```
+
+Java
+
+brave/browser/fake/android/java/src/org/chromium/brave/browser/fake/FakeServiceFactory.java
+```java
+public @Nullable FakeService getForProfile(Profile profile,
+        @Nullable ConnectionErrorHandler connectionErrorHandler) {
+    long nativeHandle = FakeServiceFactoryJni.get().getForProfile(profile);
+    MessagePipeHandle handle =
+            CoreImpl.getInstance().acquireNativeHandle(nativeHandle).
+            toMessagePipeHandle();
+    if (!handle.isValid()) {
+      return null;
+    }
+    FakeService fakeService =
+        FakeService.MANAGER.attachProxy(handle, 0);
+    if (connectionErrorHandler != null) {
+        Handler handler = ((Interface.Proxy) fakeService).getProxyHandler();
+        handler.setErrorHandler(connectionErrorHandler);
+    }
+    return fakeService;
+}
+```
+
+#### IOS
+
+IOS has separate factories because Profile subclasses `web::BrowserState` instead of `content::BrowserContext`, but the same principles apply to `ProfileKeyedServiceFactoryIOS`.
+
+#### IOS and Desktop profile checks
+
+TODO - standardize a way to maintain consistency between `GetBrowserContextToUse` methods on desktop and android. Possibly using some constants like `kRegular`, `kIncognito`, `kTor` so the same code can be used to check both `Profile` and `ProfileIOS`
+
+#### IOS keyed services and mojo
+
+obj-c
+
+brave/ios/browser/fake/fake_service_factory.mm
+```cpp
+mojo::PendingRemote<fake::mojom::FakeService> FakeServiceFactory::GetForProfile(
+    ProfileIOS* profile) {
+  auto* service = GetInstance()->GetServiceForProfileAs<fake::FakeServiceImpl>(
+      profile, true);
+  if (!service) {
+    return mojo::PendingRemote<fake::mojom::FakeService>();
+  }
+  return service->MakeRemote();
+}
+
+@implementation FakeFakeServiceFactory
++ (nullable id)getForProfile:(ProfileIOS*)profile {
+  auto* fake_service =
+      FakeServiceFactory::GetRemoteForProfile(profile);
+  if (!fake_service) {
+    return nil;
+  }
+  mojo::PendingRemote<fake::mojom::FakeService> pending_remote;
+  fake_service->Bind(pending_remote.InitWithNewPipeAndPassReceiver());
+  return [[FakeFakeServiceMojoImpl alloc]
+      initWithFakeService:std::move(pending_remote)];
+}
+```
+
+brave/ios/browser/fake/fake_service_factory_wrapper.h
+```objc
+#import <Foundation/Foundation.h>
+#include "keyed_service_factory_wrapper.h"  // NOLINT
+
+@protocol FakeFakeService;
+
+OBJC_EXPORT
+NS_SWIFT_NAME(Fake.FakeServiceFactory)
+@interface FakeFakeServiceFactory
+    : KeyedServiceFactoryWrapper < id <FakeFakeService>
+> @end
+
+#endif  // BRAVE_IOS_BROWSER_SKUS_SKUS_SDK_FACTORY_WRAPPERS_H_
+```
+
+swift
+```swift
+FakeServiceFactory.get(privateMode: privateMode),
```

Comment:
`Fake.FakeServiceFactory.get(privateMode: privateMode)`

Reply:
well, I have a question about this because code outside of components normally does not have a namespace, is there a reason we need that here?

Reply:
Depends on if you're doing an example with mojom, at which point mojom is the thing that adds the namespaces

Reply:
but the factory itself isn't mojo, that's the thing that returns the mojo interface and it would normally have a namespace since it's generally in a component

Reply:
so for instance on desktop it would normally be something like `fake::FakeService* service = FakeServiceFactory::GetInstance()->GetForProfile(profile);`

Reply:
Hmm good point, Im not sure why we bothered prefixing the Obj-C wrappers then, as of right now all of the wrappers have prefixes (`SkusSkusServiceFactory` for example), but don't necessarily need them. Perhaps its worth it to clean those up and remove the unneeded prefixes for all of the factory wrappers

Reply:
Could be due to the C++ type already using the name `FakeServiceFactory` so most likely Obj-C clobbers it? Would have to test

Reply:
It should be but I'll double check tomorrow if Obj-C and C++ types can have the same name first. If they can't we'll need some sort of prefix/namespace/alternative name for one of the two types

Reply:
@bridiver Yeah we cant have them use the same name either way: `error: redefinition of 'FakeServiceFactory' as different kind of symbol`. 

If the factory is namespaced then we're OK (i.e. `fake::FakeServiceFactory::GetInstance()->GetForProfile(profile);` but otherwise we need an alternate name for the Obj-C wrapper to ensure it uses a different name.

Reply:
Ok, I'll add the namespace back in for now then